pennylane_qulacs/qulacs_device.py

```diff
@@ -39,181 +74,239 @@
-            wires = [item for sublist in wires for item in sublist]
-        else:
-            A = self._get_operator_matrix(observable, par)
+    def apply_operations(self, operations):
+        """Apply the circuit operations to the state.
 
-        dense_gate = gate.DenseMatrix(wires, A)
-        dense_gate.update_quantum_state(self._state)
+        This method serves as an auxiliary method to :meth:`~.QulacsDevice.apply`.
 
-        expectation = inner_product(bra, self._state)
+        Args:
+            operations (List[pennylane.Operation]): operations to be applied
+        """
 
-        return expectation.real
+        for op in operations:
+            wires = op.wires
+            par = op.parameters
 
-    def probabilities(self):
-        states = itertools.product(range(2), repeat=self.num_wires)
-        probs = np.abs(self.state)**2
+            mapped_operation = self._operation_map[op.name]
+            if op.name.endswith(".inv"):
+                mapped_operation = self._get_inverse_operation(mapped_operation, wires, par)
 
-        return OrderedDict(zip(states, probs))
+            if op.name in ["QubitStateVector", "QubitStateVector.inv"]:
+                input_state = par[0]
+                input_state = _reverse_state(input_state)
 
-    def reset(self):
-        self._state.set_zero_state()
-        self._circuit = QuantumCircuit(self.num_wires)
+                if len(input_state) != 2**len(wires):
+                    raise ValueError("State vector must be of length 2**wires.")
+                if not np.isclose(np.linalg.norm(input_state, 2), 1.0, atol=tolerance):
+                    raise ValueError("Sum of amplitudes-squared does not equal one.")
+                # call qulac"s state initialization
+                self._state.load(input_state)
+
+            elif op.name in ["BasisState", "BasisState.inv"]:
+                # translate from PennyLane to Qulacs wire order
+                bits = par[0][::-1]
+                n_basis_state = len(bits)
+
+                if not set(bits).issubset({0, 1}):
+                    raise ValueError("BasisState parameter must consist of 0 or 1 integers.")
+                if n_basis_state != len(wires):
+                    raise ValueError("BasisState parameter and wires must be of equal length.")
+
+                basis_state = 0
+                for bit in bits:
+                    basis_state = (basis_state << 1) | bit
+
+                # call qulac's basis state initialization
+                self._state.set_computational_basis(basis_state)
+
+            elif op.name in ["QubitUnitary", "QubitUnitary.inv", "Hermitian", "Hermitian.inv"]:
+                if len(par[0]) != 2 ** len(wires):
+                    raise ValueError("Unitary matrix must be of shape (2**wires, 2**wires).")
+
+                if op.name in ["QubitUnitary.inv", "Hermitian.inv"]:
+                    par[0] = par[0].conj().T
```

Comment:
This isn't relevant since Hermitian will never be passed here, but the inverse of a Hermitian matrix is not the hermitian conjugate (this is only the case for a unitary matrix).